test/hotspot/jtreg/runtime/ErrorHandling/StackWalkNativeToJava.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.Utils;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/*
+ * @test StackWalkNativeToJava
+ * @bug 8316309
+ * @summary Check that walking the stack works fine when going from C++ frame to Java frame.
+ * @requires os.arch=="amd64" | os.arch=="x86_64" | os.arch=="aarch64"
+ * @requires os.family != "windows"
+ * @library /test/lib
+ * @run main/othervm StackWalkNativeToJava
+ */
+
+public class StackWalkNativeToJava {
+
+    public static void main(String[] args) throws Exception {
+        // Check stack walking works fine when sender of C++ frame
+        // is a Java native method.
+        testStackWalkNativeToJavaNative("-Xint");
+        testStackWalkNativeToJavaNative("-Xcomp", "-XX:CompileCommand=dontinline,StackWalkNativeToJava$TestNativeToJavaNative::*");
+
+        // Check stack walking works fine when sender of C++ frame
+        // is a runtime stub or interpreted Java method (VM call from Java).
+        testStackWalkNativeToJava("-Xint");
+        testStackWalkNativeToJava("-Xcomp", "-XX:TieredStopAtLevel=3",
+                                  "-XX:CompileCommand=dontinline,StackWalkNativeToJava$TestNativeToJava::*");
+    }
+
+    public static void testStackWalkNativeToJavaNative(String... extraFlags) throws Exception {
+        List<String> commands = new ArrayList<>();
+        commands.add("-Xbootclasspath/a:.");
+        commands.add("-XX:-CreateCoredumpOnCrash");
+        commands.add("-XX:+UnlockDiagnosticVMOptions");
+        commands.add("-XX:AbortVMOnException=java.lang.IllegalMonitorStateException");
+        commands.add("-XX:+ErrorFileToStdout");
+        commands.addAll(Arrays.asList(extraFlags));
+        commands.add("StackWalkNativeToJava$TestNativeToJavaNative");
+
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(commands);
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldNotContain("java.lang.RuntimeException: Reached statement after obj.wait()");
+        output.shouldNotContain("[error occurred during error reporting (printing native stack");
+        String[] res = output.getOutput().split("StackWalkNativeToJava\\$TestNativeToJavaNative\\.callNativeMethod\\(\\)V");
+        assertTrue(res.length - 1 == 2, res.length - 1);
+        output.shouldNotHaveExitValue(0);
+    }
+
+    public static class TestNativeToJavaNative {
+        public static void main(String[] args) throws Exception {
+            TestNativeToJavaNative test = new TestNativeToJavaNative();
+            test.callNativeMethod();
+        }
+
+        public void callNativeMethod() throws Exception {
+            Object obj = new Object();
+            obj.wait();
```

Comment:
Just to be clear, the aim here is to call a native method that will complete by throwing an exception, so you can abort the VM. A comment to that affect would be good. Thanks

Reply:
Done.